### PR TITLE
Column-wise covariance and correlation matrices

### LINKED ISSUE #23: Covariance/correlation matrix
I'm cleaning up stuff in CausalityTools, and came across some code I wrote for estimating the covariance/correlation matrix of a `StateSpaceSet` used for the parametric Gaussian MI/CMI estimators. The code here is significantly faster than converting to a `Matrix` and then calling `Statistics.cov`/`Statistics.cor`.

Would this be something that belongs here? 

The source code is relatively simple.

```julia
import Statistics.cov
using Statistics: mean, std
using StateSpaceSets: AbstractStateSpaceSet
using StaticArrays: @MMatrix, @MVector, SMatrix, SVector

export fastcov

# Non-allocating and faster than writing a wrapper.
# `f(x) = Statistics.cov(Matrix(x))`.
# Also accepts SubStateSpaceSets
# These functions return StaticArrays.
fastcov(x::AbstractStateSpaceSet) = fastcov(x.data)
fastmean_and_cov(x::AbstractStateSpaceSet) = fastmean_and_cov(x.data)

function fastcov(x̄, x::Vector{SVector{D, T}}) where {D, T}
    T <: AbstractFloat || error("Need `eltype(x[i]) <: AbstractFloat` ∀ i ∈ 1:length(x). Got `eltype(x[i])=$(eltype(first(x)))`")
    N = length(x) - 1
    C = @MMatrix zeros(D, D)
    x̄ = mean(x)
    Δx = @MVector zeros(D)
    @inbounds for xᵢ in x
        Δx .= xᵢ - x̄
        C .+= Δx * transpose(Δx)
    end
    C ./= N
    return SMatrix{D, D}(C)
end
# So we don't have to compute the mean twice at every iteration.
function fastcov(x::Vector{SVector{D, T}}) where {D, T}
    T <: AbstractFloat || error("Need `eltype(x[i]) <: AbstractFloat` ∀ i ∈ 1:length(x). Got `eltype(x[i])=$(eltype(first(x)))`")

    μ = mean(x)
    fastcov(μ, x)
end
function fastmean_and_cov(x::Vector{SVector{D, T}}) where {D, T}
    μ = mean(x)
    Σ = fastcov(μ, x)
    return μ, Σ
end

# Non-allocating and faster than writing a wrapper.
export fastcor
fastcor(x::AbstractStateSpaceSet) = fastcor(x.data)
function fastcor(x::Vector{SVector{D, T}}) where {D, T}
    N = length(x)
    μ, Σ = fastmean_and_cov(x)
    σ = std(x)
    C = @MMatrix zeros(D, D)
    for j in 1:D
        for i in 1:D
            C[i, j] = Σ[i, j] / (σ[i] * σ[j])
        end
    end
    return SMatrix{D, D}(C)
end
```

Verifying that the implementations are actually correct:

```julia
julia> using Test

julia> using Statistics

julia> D = StateSpaceSet(rand(100, 3));

julia> @test all(fastcov(D) .≈ Statistics.cov(Matrix(D)))
Test Passed

julia> @test all(fastcor(D) .≈ Statistics.cor(Matrix(D)))
Test Passed
```

## Timing

```julia
julia> D = StateSpaceSet(rand(10000, 5)); 

julia> mcor(x) = Statistics.cor(Matrix(x))
mcor (generic function with 1 method)

julia> fastcor(D); @btime fastcor($D);
  59.708 μs (0 allocations: 0 bytes)

julia> mcor(D); @btime mcor($D);
  89.333 μs (13 allocations: 782.31 KiB)
```

### PR DESCRIPTION
**Pull request by kahaaga**
- Adds `cov`, `cor` and `mean_and_cov` functions. 
- Tests correctness by comparing with `Statistics.cov`/`Statistics.cor`.
- Adds a section "Basic statistics" in the docs, where `standardize` is also moved.
- Increments version to `1.5` and updates change log.

Fixes #23.

### CODE REVIEW

**Comment by kahaaga** [comment]
Should I pin Documenter to 0.27 here too, so the docs can build?  (ref https://github.com/JuliaDynamics/CausalityTools.jl/issues/358)

**Comment by Datseris** [comment]
I don't get the documenter problem. The doctheme pins documenter by itself:

https://github.com/JuliaDynamics/doctheme/blob/master/build_docs_with_style.jl#L4

File: src/statespaceset.jl [hunk 1]
```diff
@@ -262,3 +264,76 @@ function standardized_timeseries(d::AbstractStateSpaceSet)
...
+
+Compute the covariance matrix `m` from the columns of `d`, where `m[i, j]` is the covariance
+between `d[:, i]` and `d[:, j]`.
+"""
+cov(x::AbstractStateSpaceSet) = fastcov(x.data)
+
+"""
+    mean_and_cov(d::StateSpaceSet) → μ, m::SMatrix
+
+Returns a tuple of the column means `μ` and covariance matrix `m`.
```
Comment (Datseris) [inline src/statespaceset.jl]:
```suggestion
Return a tuple of the column means `μ` and covariance matrix `m`. 
```

File: src/statespaceset.jl [hunk 2]
```diff
@@ -262,3 +264,76 @@ function standardized_timeseries(d::AbstractStateSpaceSet)
...
+"""
+mean_and_cov(x::AbstractStateSpaceSet) = fastmean_and_cov(x.data)
+
+"""
+    cor(d::StateSpaceSet) → m::SMatrix
+
+Compute the corrlation matrix `m` from the columns of `d`, where `m[i, j]` is the 
+correlation between `d[:, i]` and `d[:, j]`.
+"""
+cor(x::AbstractStateSpaceSet) = fastcor(x.data)
```
Comment (Datseris) [inline src/statespaceset.jl]: What's the difdference between correlation and covariance? AAren't they the same thing up to a normalization factor? We should say this.

**Comment by kahaaga** [comment]
Skjermbilde 2023-10-14 kl  13 03 30

The CI doc build fails, though. So we can't inspect any preview docs, because they're not being pushed.

**Comment by Datseris** [comment]
Yes, I know the docbuilds fail but I can't understand why documenter v1 is being installed. In any case, pinning it _here_ to v1 is not an option, as then I would have to undo the pin in the 10+ packages that compose DynamicalSystems.jl.

The only solution is to solve https://github.com/JuliaDynamics/doctheme/issues/2  as far as the docs are concerned.

File: src/statespaceset.jl [hunk 3]
```diff
@@ -262,3 +264,76 @@ function standardized_timeseries(d::AbstractStateSpaceSet)
...
+using Statistics: mean, std
+using StaticArraysCore: MMatrix, MVector, SMatrix, SVector
+
+"""
+    cov(d::StateSpaceSet) → m::SMatrix
+
+Compute the covariance matrix `m` from the columns of `d`, where `m[i, j]` is the covariance
+between `d[:, i]` and `d[:, j]`.
+"""
+cov(x::AbstractStateSpaceSet) = fastcov(x.data)
```
Comment (Datseris) [inline src/statespaceset.jl]: we should also replace all instances of `x.data` here with `vec(x)` that does the same thing but on the basis of functions.

**Comment by Datseris** [comment]
@kahaaga shall we finish this? I think you replace `x.data` with `vec(x)` and its done.

**Comment by Datseris** [comment]
docs have been fixed for a while now.

**Comment by kahaaga** [comment]
Do we keep the Julia version at 1.5? The LTS is 1.6 at the moment

**Comment by kahaaga** [comment]
As for the distinction between correlation and covariance, I don't think it's necessary to specify here. This is common knowledge. The Statistics.jl package also doesn't give any details on the relationship between the two quantities.

**Comment by kahaaga** [comment]
I am not sure whether the docs fail now because of a repo-intrinsic issue, or if it is due to the organisation wide repo build. Maybe you know more, @Datseris ?

**Comment by Datseris** [comment]
i'll fix it...

**Pull request merged**